### PR TITLE
Created PR template for the rccl repo

### PR DESCRIPTION
**Pull request by corey-derochie-amd**
This template will auto-populate the description box on all future PR's in this repo. It will appear as follows:

## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** _"Internal", or link to GitHub issue (if applicable)._

**What were the changes?**  
_One sentence describing the work done._

**Why were the changes made?**  
_Explain the motivation behind the work. Provide any publicly-available historical context._

**How was the outcome achieved?**  
_Technical details behind the work. Explain any publicly-available hardware peculiarities._

**Additional Documentation:**  
_What else should the reviewer know?_

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:5]: I would suggest "PR Buddy" being automatic: the person responds to the review will be assumed as buddy

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:11]: Customer specific info should not be put into PR. We can put a link to JIRA

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:14]: AMD hardware gotchas cannot be put in PR. We can put a link to JIRA

**Review by wenkaidu — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:5]: I don't believe this can be automated. It might be best if whoever becomes the buddy inserts their own handle here?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:11]: Agreed. I forgot to remove this line.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: I don't think hardware gotchas are necessarily proprietary info?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
_Explain the motivation behind the work. Provide any publicly-available historical context._
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: If AMD has published an errata, then yes, we can put a link to the errata. Otherwise, I don't think we should mention anything in public PR

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: Some PRs fix customer platform that we may not have access. There needs some flexibility here

**Review by wenkaidu — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:5]: I think as soon as someone responds, the large reviewers list is removed. You can see the example in this PR. Right now only my name is up there

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md]: Eed-to-end ML workload test steps can be very complicated. Again maybe better use link to JIRA, which is supposed to have detailed reproduction steps.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: I think in this case, "gotchas" means "particularities", not necessarily errata. Perhaps I should reword?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: As a general comment: if this cannot be AGS wide requirements, at least, this needs to be reviewed and adopted in mathlib organization.

**Review by wenkaidu — CHANGES_REQUESTED** [review]

**Comment by corey-derochie-amd** [comment]
PR should only contain changes to solve a single task or problem. Multiple problems require multiple PR's.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md]: Test functionality and performance is too much burden for "PR Buddy". This task should be done by CI steps.

**Comment by corey-derochie-amd** [comment]
Format the PR title as follows:
[MODULE] Short description of feature

tbc...

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: It doesn't seem right for every PR to update CHANGELOG, which should be a summary for each release.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: This needs to be moved to before merging

**Comment by corey-derochie-amd** [comment]
Add guidelines around how long a PR should live. 2 days for now.

**Comment by wenkaidu** [comment]
> Add guidelines around how long a PR should live. 2 days for now.
This is not common practice. A PR can take a long time to get reviewed and merged. Closing a PR should have a valid reason. According to documentation on github itself:
You may choose to close a pull request without [merging it into the upstream branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request). This can be handy if the changes proposed in the branch are no longer needed, or if another solution has been proposed in another branch.

**Comment by corey-derochie-amd** [comment]
> This is not common practice. A PR can take a long time to get reviewed and merged. Closing a PR should have a valid reason. According to documentation on github itself:
> You may choose to close a pull request without [merging it into the upstream branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request). This can be handy if the changes proposed in the branch are no longer needed, or if another solution has been proposed in another branch.

Sorry, this was meant to be a _minimum_ amount of time open, to allow opportunity for other maintainers to review if they want to before the buddy approves and it is merged. The timeline still warrants discussion though.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:5]: If a second maintainer comments on the review, are they also added to the reviewers list, or just the first?

Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:5]: Never mind. It appears to be inconsistent. Other PRs have all the people back on again even there were some reviews. Anyway, I don't think introducing "PR buddy" is correct workflow. There are 3 roles for a PR: author(s), reviewer(s) and tester(s). CI is one of the testers. We should not expect reviewers assume new "buddy" role to test PRs. That will be too much burden for reviewers. Test may need dedicated setup and complicated procedures.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:14]: If that hardware "particularities" is already in public documentation, then yes.

**Review by wenkaidu — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: The entire approval checklist is meant to be performed by the maintainers before merging. Perhaps there should be a pre-merge and post-merge checklist?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:5]: I will shift this away from the role of "PR buddy" and toward defining the responsibilities of the approver and tester.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: The direction here is to be more meticulous. Ensure that we document all relevant test cases, and also document any of those that were skipped, and explain why.

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: If a customer platform is involved, do not mention in the PR but document internally what the system was and how it was tested upon.

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: End-to-end is out of scope for this stage of development.

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: Rather than burden with the reviewer with testing functionality and performance, they must ensure that those steps have been covered.

**Comment by BertanDogancay** [comment]
Few opinions after reading through this PR:

1) It would be nice to have a drop box that lists all types of architectures including an option 'all' so that author can select the architectures that the PR will affect. Sometimes our bug fixes only apply to certain archs.

2) In my opinion, one text box called description is enough. Again, the author of the PR should be responsible from what information is needed or doesn't need to be included in the PR. 

3) Instead of module prefixes for PRs as previously proposed like the name of the file/module, it might be better to list categories in the template where the author is responsible from selecting the right one and include it as prefix to the PR title. (Ex. bug, build, refactor, sync, doc, feature, test etc...)

@corey-derochie-amd  @wenkaidu

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:5]: Since there are so many SoCs, platforms and use cases, unless some exceptional cases, CI should be the only tester. It will be a huge waste of resources to require anyone to manually execute these huge list of testing. We should spend effort to enhance CI testing instead. Engineering resource should be used to improve and debugg CI testing failures.

**Comment by wenkaidu** [comment]
> > This is not common practice. A PR can take a long time to get reviewed and merged. Closing a PR should have a valid reason. According to documentation on github itself:
> > You may choose to close a pull request without [merging it into the upstream branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request). This can be handy if the changes proposed in the branch are no longer needed, or if another solution has been proposed in another branch.
> 
> Sorry, this was meant to be a _minimum_ amount of time open, to allow opportunity for other maintainers to review if they want to before the buddy approves and it is merged. The timeline still warrants discussion though.

I am not sure the value added by delaying. PR is not wine which quality improves with aging. Author of a PR should not be punished by someone else's failure to promptly review the PR.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md]: maybe approval checklist and merging checklist?
We normally always leave merging to author as he/she has most updated knowledge on a PR.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: Anything that impacts a customer or another team should be logged. I'll reword to leave this up to the discretion of the maintainer.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: So far, the only item on the merge checklist is to wait for approval, so I'm going to leave that off for now. All approval steps will be intended for before merging.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: Not sure how reviewer can "Verify appropriate developer testing was done"? Email author? I think this needs to be moved to checklist before merging. Author is responsible for this testing

**Comment by corey-derochie-amd** [comment]
> > > This is not common practice. A PR can take a long time to get reviewed and merged. Closing a PR should have a valid reason. According to documentation on github itself:
> > > You may choose to close a pull request without [merging it into the upstream branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request). This can be handy if the changes proposed in the branch are no longer needed, or if another solution has been proposed in another branch.
> > 
> > 
> > Sorry, this was meant to be a _minimum_ amount of time open, to allow opportunity for other maintainers to review if they want to before the buddy approves and it is merged. The timeline still warrants discussion though.
> 
> I am not sure the value added by delaying. PR is not wine which quality improves with aging. Author of a PR should not be punished by someone else's failure to promptly review the PR.

I'm leaving this out for now but there is concern about if a PR gets approved before another maintainer can review it who has good feedback. I think it'll be up to the maintainers to coordinate to not approve too quickly or too slowly.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: I think this also needs to be moved to checklist before merging. Author is responsible for confirming CI test results before merging

**Review by wenkaidu — CHANGES_REQUESTED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: I think I see your point now. I will adjust it.

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: Actually, for this item, the idea is that the author should not have submitted a PR on untested work, so the approver should not approve unless that happened. The author is responsible for the testing but the approver is responsible not to approve unless the testing was done.

The author should be listing their test cases in the Details section, which is already covered by "Ensure above details are appropriately filled", so I will remove this.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: I think we need some clarifications on when to update CHANGELOG

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: Agreed. My directive so far is that:
- Any changes that impact the customer should be logged.
- Any changes that impact any other team that is dependent on us should be logged.

Do you have any suggestions to further refine that?

Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md]: The directive looks good to me. Can you add them to template?

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: I've added those and reworded to make them relevant to external authors and readers.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: This needs more clarification: what is internal document referred here?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
```

**Comment by corey-derochie-amd** [comment]
> Few opinions after reading through this PR:
> 
> 1. It would be nice to have a drop box that lists all types of architectures including an option 'all' so that author can select the architectures that the PR will affect. Sometimes our bug fixes only apply to certain archs.
> 2. In my opinion, one text box called description is enough. Again, the author of the PR should be responsible from what information is needed or doesn't need to be included in the PR.
> 3. Instead of module prefixes for PRs as previously proposed like the name of the file/module, it might be better to list categories in the template where the author is responsible from selecting the right one and include it as prefix to the PR title. (Ex. bug, build, refactor, sync, doc, feature, test etc...)
> 
> @corey-derochie-amd @wenkaidu

I don't believe the drop box feature is available in markdown. The other suggestions are outside the scope of this template so we won't worry about them for now.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: This has been addressed by removing it.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: We can propagate this to the org once we've used it and iterated on it. Let's revisit this later.

**Review by wenkaidu — APPROVED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:2]: actually we will need to link to internal JIRA. Otherwise, we lost most context on many PRs

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:16]: I suggest this field is used to link to JIRAs or github issues

**Review by wenkaidu — DISMISSED** [review]

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:2]: We have guidance from legal not to refer to JIRA tickets in the public repo, even though it doesn't leak non-public info. I think what we want to do is link to the PR from JIRA but not the other way around.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md]: For bug fix PRs, test cases are not straightforward most of time, i.e. need framework dockers, special test procedure and exact gfx/machines, etc. Again, better reference to JIRA or github issues for details

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (gilbertlee-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:2]: Perhaps just the name of the internal JIRA is sufficient.  Not an actual hyperlink.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (gilbertlee-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: I too would recommend removing this section.

Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: If an external contributor is submitting a PR, they won't have a JIRA ticket to reference. They may have a Github issue, so perhaps it would be possible to offload this type of documentation there in that case.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:2]: sure...although I am not clear if it makes any difference to users not on VPN

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:20]: Let's add if there is NCCL API version change. It looks like version in current CHANGELOG was stuck in 2.16

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (wenkaidu) [inline .github/PULL_REQUEST_TEMPLATE.md:15]: Can we just make the title to "** Link to github issues or internal JIRA **"? So there is no confusion what should be "documentation" or "additional information".

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (yhuiYH) [reply .github/PULL_REQUEST_TEMPLATE.md:2]: I agree with Gilbert.  No hyperlinks to anything internal.  
You may mention a JIRA number (at most), but no hyperlinks.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (yhuiYH) [reply .github/PULL_REQUEST_TEMPLATE.md:15]: ** Link to GitHub issue# (if applicable)** is fine, do not mention JIRA.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (yhuiYH) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: No JIRA.

Comment (wenkaidu) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: Reason? JIRA is just a string like SWDEV-416920. What is the concern here?

Comment (yhuiYH) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: Your comment above says "link to JIRA".  No links to JIRAs should be provided.  Links to GitHub issues are ok.

Your string example is ok.  but be careful.  As people are in a rush to fill in details in PR, they may copy/paste the internal links, which is NOT good.  To avoid any mistakes, a "SAFE" way, is to avoid mentions to JIRA completely.  

Err on the side of caution, and avoid JIRA is best practice.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** _"Internal", or link to GitHub issue (if applicable)._

**What were the changes?**  
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:2]: @yhuiYH Elsewhere, you say not to mention JIRA. Can you clarify whether or not a JIRA number is acceptable here?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:20]:
```suggestion
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
```

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:2]: I've addressed this elsewhere.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:15]: I've addressed this elsewhere without mention of JIRA.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md:16]: I've addressed this elsewhere.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (corey-derochie-amd) [reply .github/PULL_REQUEST_TEMPLATE.md]: Removing this.

Comment (corey-derochie-amd) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
```

**Review by gilbertlee-amd — APPROVED** [review]

**Review by wenkaidu — APPROVED** [review]

**Pull request merged**
